Fix Vault No Economy Plugin Found Error

DerMistkaefer opened the pull request:
Fix the error if the user use vault but have no economy plugin. And then disable the plugin peacefully.

Current Error:
```
[18:24:15 ERROR]: Error occurred while enabling ChestShop v3.12.3-SNAPSHOT (build 423) (Is it up to date?)
java.lang.NullPointerException: Cannot invoke "net.milkbowl.vault.economy.Economy.getName()" because "com.Acrobot.ChestShop.Listeners.Economy.Plugins.VaultListener.provider" is null
        at ChestShop.jar/com.Acrobot.ChestShop.Listeners.Economy.Plugins.VaultListener.getProviderInfo(VaultListener.java:68) ~[ChestShop.jar:?]
        at ChestShop.jar/com.Acrobot.ChestShop.Dependencies.loadEconomy(Dependencies.java:114) ~[ChestShop.jar:?]
        at ChestShop.jar/com.Acrobot.ChestShop.Dependencies.loadPlugins(Dependencies.java:80) ~[ChestShop.jar:?]
        at ChestShop.jar/com.Acrobot.ChestShop.ChestShop.onEnable(ChestShop.java:163) ~[ChestShop.jar:?]
        at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:288) ~[paper-mojangapi-1.21.1-R0.1-SNAPSHOT.jar:?]
        at io.papermc.paper.plugin.manager.PaperPluginInstanceManager.enablePlugin(PaperPluginInstanceManager.java:202) ~[server.jar:1.21.1-DEV-b48403b]
        at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.enablePlugin(PaperPluginManagerImpl.java:109) ~[server.jar:1.21.1-DEV-b48403b]
        at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:520) ~[paper-mojangapi-1.21.1-R0.1-SNAPSHOT.jar:?]
        at realms-spigot-core-plugin-0.1.jar/de.cytooxien.realms.spigot.core.plugin.RealmPluginManager.enablePlugin(RealmPluginManager.java:537) ~[realms-spigot-core-plugin-0.1.jar:?]
        at realms-spigot-core-plugin-0.1.jar/de.cytooxien.realms.spigot.core.RealmsPluginsLoader.loadUserPlugins(RealmsPluginsLoader.java:74) ~[realms-spigot-core-plugin-0.1.jar:?]
        at realms-spigot-core-plugin-0.1.jar/de.cytooxien.realms.spigot.core.local.RealmPostInitializationUseCase.lambda$loadUserPlugins$2(RealmPostInitializationUseCase.java:51) ~[realms-spigot-core-plugin-0.1.jar:?]
        at realms-spigot-core-plugin-0.1.jar/de.cytooxien.realms.spigot.core.util.RunnerImpl.run(RunnerImpl.java:11) ~[realms-spigot-core-plugin-0.1.jar:?]
        at realms-spigot-core-plugin-0.1.jar/de.cytooxien.realms.spigot.core.util.Runner.either(Runner.java:9) ~[realms-spigot-core-plugin-0.1.jar:?]
        at realms-spigot-core-plugin-0.1.jar/de.cytooxien.realms.spigot.core.local.RealmPostInitializationUseCase.loadUserPlugins(RealmPostInitializationUseCase.java:50) ~[realms-spigot-core-plugin-0.1.jar:?]
        at realms-spigot-core-plugin-0.1.jar/de.cytooxien.realms.spigot.core.local.RealmPostInitializationUseCase.afterFetchRealmInfo(RealmPostInitializationUseCase.java:30) ~[realms-spigot-core-plugin-0.1.jar:?]
        at realms-spigot-core-plugin-0.1.jar/de.cytooxien.realms.spigot.core.RealmsCorePlugin.lambda$fetchRealmInfo$1(RealmsCorePlugin.java:161) ~[realms-spigot-core-plugin-0.1.jar:?]
        at org.bukkit.craftbukkit.v1_21_R1.scheduler.CraftTask.run(CraftTask.java:86) ~[server.jar:1.21.1-DEV-b48403b]
        at org.bukkit.craftbukkit.v1_21_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:475) ~[server.jar:1.21.1-DEV-b48403b]
        at net.minecraft.server.MinecraftServer.c(MinecraftServer.java:1724) ~[server.jar:1.21.1-DEV-b48403b]
        at net.minecraft.server.dedicated.DedicatedServer.c(DedicatedServer.java:473) ~[server.jar:1.21.1-DEV-b48403b]
        at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:1596) ~[server.jar:1.21.1-DEV-b48403b]
        at net.minecraft.server.MinecraftServer.y(MinecraftServer.java:1302) ~[server.jar:1.21.1-DEV-b48403b]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:329) ~[server.jar:1.21.1-DEV-b48403b]
        at java.base/java.lang.Thread.run(Unknown Source) [?:?]
```

After PR:
```
[20:56:10 ERROR]: [ChestShop] No Vault compatible Economy plugin found!
[20:56:10 INFO]: [ChestShop] Disabling ChestShop v3.12.3-SNAPSHOT (compiled at 2024-12-09T19:55:26Z)
```

Phoenix616 reviewed (changes requested):
This will not solve the issue as economy plugins can enable after ChestShop so ChestShop should not disable at all, the checkSetup method itself seems not really match that design anymore either.

The better solution for the bug would probably be that the economy plugin metric needs to be registered/updated when the Economy provider changes.

DerMistkaefer commented:
Okay, I’ve incorporated your requested changes. The issue was already present before: if the economy plugin is initialized at a later point, the exception would still be thrown. 

Now I’ve handled everything through a listener that reacts to provider changes. However, I’m not entirely sure if updating through bStats works as intended, since I have no prior experience with it.

I’ve also changed the economy plugin existence check to the ServerLoadEvent, which is executed once all plugins have finished loading.

Phoenix616 commented on src/main/java/com/Acrobot/ChestShop/Dependencies.java:
```diff
@@ -111,7 +111,7 @@ private static boolean loadEconomy() {
 
         ChestShop.getMetrics().addCustomChart(ChestShop.createStaticDrilldownStat("economyAdapter", plugin, Bukkit.getPluginManager().getPlugin(plugin).getDescription().getVersion()));
-        ChestShop.getMetrics().addCustomChart(ChestShop.createStaticDrilldownStat("economyPlugin", economy.getProviderInfo().getName(), economy.getProviderInfo().getVersion()));
+        economy.registerProviderChangeListener((EconomyAdapter.ProviderInfo provider) -> ChestShop.getMetrics().addCustomChart(ChestShop.createStaticDrilldownStat("economyPlugin", provider.getName(), provider.getVersion())));
```
Custom charts are a set in the metrics class so this would lead to a new chart getting registered every time the provider changes. I'm unsure if this actually causes issues when submitting it but I feel like we should have a better approach to avoid that. It should be as simple as just checking if `economy#getProviderInfo` returns `null`when the `providingPlugin`tbh.

DerMistkaefer replied:
So you're suggesting that we should update the chart whenever the provider changes. Now we have this problem.
Is it possible to set the metric once within the ServerLoadEvent?

Phoenix616 replied:
Ideally you would just return the info of the currently used provider whenever the metrics are submitted as the charts can take a Callable which queries the current value. This wasn't done for the economy plugin before though so there is no method for that, but you should be able to easily adjust the implementation of `ChestShop#createStaticDrilldownStat` to achieve that.

Only setting it on the ServerLoadEvent would make the metric wrong as it wouldn't use the current provider (although that is already what's happening right now but the best way to fix this issue imo)

Phoenix616 commented on src/main/java/com/Acrobot/ChestShop/ChestShop.java:
```diff
@@ -536,6 +539,13 @@ public static DrilldownPie createStaticDrilldownStat(String statId, String value
+    public static DrilldownPie createStaticDrilldownStat(String statId, Callable<EconomyAdapter.ProviderInfo> callableProviderInfo) {
+        return new DrilldownPie(statId, () -> {
+            EconomyAdapter.ProviderInfo providerInfo = callableProviderInfo.call();
+            return ImmutableMap.of(providerInfo.getName(), ImmutableMap.of(providerInfo.getVersion(), 1));
```
This will throw an NPE when no provide is set. I realize that this shouldn't be happening as the plugin disables on load but you made getProviderInfo Nullable so it should be checked here.

Phoenix616 commented on src/main/java/com/Acrobot/ChestShop/ChestShop.java:
```diff
@@ -51,6 +52,7 @@
 import com.google.common.io.ByteArrayDataOutput;
 import com.google.common.io.ByteStreams;
 
+import jdk.internal.joptsimple.util.KeyValuePair;
```
This seems unused.

DerMistkaefer replied:
Upps....

Phoenix616 reviewed (commented):
Looks good! :) Just the last two things on my comments.

Phoenix616 reviewed (approved):
Looks good now 👍 Thank you very much!
